src/stan/services/pathfinder/multi.hpp

```diff
@@ -105,136 +109,214 @@ inline int pathfinder_lbfgs_multi(
-              logger.error(std::string("Pathfinder iteration: ")
-                           + std::to_string(iter) + " failed.");
-              return;
+            if (psis_resample && calculate_lp) {
+              auto toss_write = [](auto&&... /* x */) {};
+              auto pathfinder_ret
+                  = stan::services::pathfinder::pathfinder_lbfgs_single<true>(
+                      model, *(init[iter]), random_seed, stride_id + iter,
+                      init_radius, history_size, init_alpha, tol_obj,
+                      tol_rel_obj, tol_grad, tol_rel_grad, tol_param,
+                      num_iterations, num_elbo_draws, num_draws,
+                      save_iterations, refresh, interrupt, logger,
+                      init_writers[iter], toss_write,
+                      single_path_diagnostic_writer[iter], calculate_lp,
+                      psis_resample);
+              if (unlikely(std::get<0>(pathfinder_ret) != error_codes::OK)) {
+                logger.error(std::string("Pathfinder iteration: ")
+                             + std::to_string(iter) + " failed.");
+                return;
+              }
+              lp_calls += std::get<2>(pathfinder_ret);
+              elbo_estimates.push_back(
+                  std::make_pair(iter, std::move(std::get<1>(pathfinder_ret))));
+            } else {
+              // For no psis, have single write to both single and multi writers
+              using multi_writer_t = stan::callbacks::multi_writer<
+                  SingleParamWriter,
+                  stan::callbacks::concurrent_writer<ParamWriter>>;
+              multi_writer_t multi_param_writer(
+                  single_path_parameter_writer[iter], safe_write);
+              auto pathfinder_ret
+                  = stan::services::pathfinder::pathfinder_lbfgs_single<false>(
+                      model, *(init[iter]), random_seed, stride_id + iter,
+                      init_radius, history_size, init_alpha, tol_obj,
+                      tol_rel_obj, tol_grad, tol_rel_grad, tol_param,
+                      num_iterations, num_elbo_draws, num_draws,
+                      save_iterations, refresh, interrupt, logger,
+                      init_writers[iter], multi_param_writer,
+                      single_path_diagnostic_writer[iter], calculate_lp,
+                      psis_resample);
+              if (pathfinder_ret.first != error_codes::OK) {
+                logger.error(std::string("Pathfinder iteration: ")
+                             + std::to_string(iter) + " failed.");
+                return;
+              }
+              lp_calls += pathfinder_ret.second;
             }
-            individual_lp_ratios[iter] = std::move(std::get<1>(pathfinder_ret));
-            individual_samples[iter] = std::move(std::get<2>(pathfinder_ret));
-            lp_calls += std::get<3>(pathfinder_ret);
           }
         });
+    safe_write.wait();
```

Comment:
Q: Rather than needing to call this, should the concurrent_writer type have a destructor that blocks?

Reply:
I like being explicit about it but I can also just make a destructor